src/util/uriKeyParams.ts

```diff
@@ -0,0 +1,20 @@
+export function getUriKeyParams(uri: string): string[] {
+  const params = uri.match(/%\{(\w|\d)+\}/g)
+  return params == null ? [] : params
+}
+
+export function replaceKeyParams<T extends string>(
+  uri: string,
+  options: { [P in T]?: string }
+): string {
+  const keyParams = getUriKeyParams(uri)
```

Comment:
The `getUriKeyParams` should just be inlined as `const keyParams = uri.match(/%\{(\w|\d)+\}/g) ?? []`